### DISCUSSION TITLE
How to change the pieceSize and redraw?

### DISCUSSION THREAD

**Discussion by sebastiaandegeus** [flbulgarelli/headbreaker#59] [Q&A]
Hi,

I would like to add 2 buttons to increase and decrease the pieceSize and redraw but it is not working.

**Comment by flbulgarelli** [comment]
Hi @sebastiaandegeus!

What have you tried? Could you share some code, please?

Regards

**Comment by sebastiaandegeus** [comment]
Hi, 

Something like this 

```html
<script src="https://flbulgarelli.github.io/headbreaker/js/headbreaker.js"></script>

<body>
  <button id="increase">
    increase
  </button>

  <button id="decrease">
    decrease
  </button>
  <div id="puzzle"></div>
  <script>
    const autogen = new headbreaker.Canvas('puzzle', {
      width: 800,
      height: 650,
      pieceSize: 100,
      proximity: 20,
      borderFill: 10,
      strokeWidth: 2,
      lineSoftness: 0.18,
    });
    autogen.autogenerate({
      horizontalPiecesCount: 2,
      verticalPiecesCount: 2,
      metadata: [{
          color: '#B83361'
        },
        {
          color: '#B87D32'
        },
        {
          color: '#A4C234'
        },
        {
          color: '#37AB8C'
        }
      ]
    });
    autogen.draw();

    document.getElementById("increase").addEventListener('click', function() {
      console.log("increase")
      autogen.pieceSize = autogen.pieceSize + 10;
      autogen.redraw();
    })

    document.getElementById("decrease").addEventListener('click', function() {
      console.log("decrease")
      autogen.pieceSize = autogen.pieceSize - 10;
      autogen.redraw();
    })

  </script>
</body>
```

**Reply by sebastiaandegeus** [reply]
[Example](https://jsfiddle.net/L6qjathd/1/)

**Comment by sebastiaandegeus** [comment]
Redraw does not seem to update the canvas. 

I console logged the canvas.puzzle afterwards and size or number of pieces is updated but the canvas does not change it. It does change the rendering of the image though so something is happening. 

Any idea what could be wrong?

**Comment by flbulgarelli** [comment]
Hi! It looks like there are two different issues here: 

1. `pieceSize` is not a scalar, but a `size` object, so updating it by using numeric operations like `+=` won't work. You must use operations like `headbreaker.diameter` to create new sizes.  
2. `redraw` is meant to refresh the canvas after `Puzzle` model changes, but it won't work with deep structural changes to the canvas itself. You will need to re-create the `canvas` and the `div` that contains it by hand. Please have a look at this question, where the author also needed to perform a similar structural change to the canvas dynamically https://github.com/flbulgarelli/headbreaker/issues/24. 

Hope it helps!

**Comment by flbulgarelli** [comment]
PS: Have you tried scaling the canvas - using the `scale` method - instead of changing the pieces sizes? I think that may be a better solution to your actual problem - changing the "zoom" of the visual representation instead of modifying the actual structure. Here is a working example based on your code: https://codepen.io/flbulgarelli/pen/gOBdvyR

**Comment by sebastiaandegeus** [comment]
Hi, I tried this scale solution and it works. The only issue I have is that the shuffle area becomes smaller and if I shuffle the area where the piece are put has also been scaled down. 

![image](https://github.com/flbulgarelli/headbreaker/assets/1560127/dc2f6da4-11ea-4ec4-9e69-93d82b463734)

![image](https://github.com/flbulgarelli/headbreaker/assets/1560127/a299fdb6-881e-4985-86f5-ccb303229d6a)

Is there a way to allow the shuffle method to use the entire canvas again?
